### ISSUE TITLE
Global state in SystemC

### ISSUE THREAD

**Issue by nmeum** [accellera-official/systemc#8]
I am currently working on a specialized SystemC project where I want to restart the SystemC simulation process multiple times. I have a working proof of concept which does this by providing a custom `main` function in which I reset the global SystemC simulation context (`sc_core::sc_curr_simcontext`) and call `sc_core::sc_elab_and_sim` in a loop. As an example, consider the following code:

```c
int main(int argc, char **argv) {
        // …

	for (i = 0; !exit_simulation; i++) {
		printf("\n##\n# %zuth execution\n##\n", i);

		// Reset SystemC simulation context
		if (sc_core::sc_curr_simcontext)
			delete sc_core::sc_curr_simcontext;
		sc_core::sc_curr_simcontext = NULL;

		if ((ret = sc_core::sc_elab_and_sim(argc, argv)))
			return ret;
	}

        // …
}
```

I am aware that this code relies on specific implementation details of the `sc_get_curr_simcontext()` function and is thus not conforming to IEEE Std 1666-2001. In fact, the code above only works because `sc_get_curr_simcontext()` currently creates a new simulation context if the global `sc_core::sc_curr_simcontext` is NULL:

https://github.com/accellera-official/systemc/blob/c81e0d9898c5a78b63418502b6687d6a3efb13b4/src/sysc/kernel/sc_simcontext.h#L420-L423

Apart from relying on implementation-defined behaviour, an additional problem with the code above is that SystemC contains global variables which are independent of the simulation context. Some of these variables even hold references to the simulation context which cannot be updated and thus cause a double free on program termination.  One example of such variables is the `sc_event::none` variable which was added in 55da81d96e73ea8bccad1dbab59c1bae92cee661:

https://github.com/accellera-official/systemc/blob/55da81d96e73ea8bccad1dbab59c1bae92cee661/src/sysc/kernel/sc_event.h#L297-L298

As `sc_event::none` is a global variable, it is initialized on program startup and holds a reference to a simulation context (created through the invocation of `sc_get_curr_simcontext()` by the `sc_event` constructor). This reference is not updated by the loop from above, in fact in cannot be updated as the `m_simc` member of `sc_event` is private. An additional global variable with the same problem is `sc_process_handle::non_event`:

https://github.com/accellera-official/systemc/blob/f567c9c40c08f5aaee47529802bf598b8472ce4a/src/sysc/kernel/sc_process.cpp#L50

I am currently considering working on a modified version of SystemC which addresses issue related to such global variables. I created this issue to figure out if there is an interest in reducing global state in SystemC and if patches in this regard would be accepted.

**Comment by mattgately** [comment]
I agree that this is a useful feature request.  I am also doing something where I reset the simcontext to enable rerunning of the simulation kernel.

**Comment by wltr** [comment]
Has there be any progress on this? Not being able to restart the simulation is the biggest limitation of SystemC imho and I'd be very interested in this.

**Comment by nmeum** [comment]
We are using the patches from #9 as part of a [SystemC-based symbolic execution engine](https://github.com/agra-uni-bremen/symex-vp), there is some room for improvement but the general idea does seem to work.

**Comment by AndrewGoodrich** [comment]
There are number of global structures that really should be reset when trying to use a "fresh" simulation context. I think we might be better off if we had a reinitialize() method in the sc_simcontext class, that would allow the code to properly reinitialize those global structures.

**Comment by thatTPR** [comment]
Is this solved? If not I would appreciate pointers on helping make this feature available.

**Comment by maehne** [comment]
@AndrewGoodrich: Might you be able to provide some pointers to @thatTPR?

**Comment by AndrewGoodrich** [comment]
The original mention of the issue is in #8 in accelera-official. It is issue #435 in OSCI-wg/systemc. accelera-official issue #20 describes what is in internal #435. I closed #20 when I created #435. I have code with the change present, and will generate a pull request in the near future.

Andy

**Comment by wz108578656** [comment]
So does this issue haave a proper solution for now? If yes, many appriciate for anybody to show an example on how to multiple restart the simulation.

**Comment by maehne** [comment]
@AndrewGoodrich: Could you comment on the state in the released SystemC 2.3.4? As far as I can see from its release notes, your mentioned improvements are not yet part of it.

**Comment by AndrewGoodrich** [comment]
The change for this was put into the master branch and the issue #435 closed on 14 December 2021.

If sc_event::none() appears in sysc/communication/sc_interface.cpp the changes are present.

Andy

**Comment by AmeyaVS** [comment]
I don't these changes in this repo here:
https://github.com/accellera-official/systemc/blob/39740075f47fedbce242808d357fcfb6d3d33957/src/sysc/communication/sc_interface.cpp#L55

**Comment by AmeyaVS** [comment]
Seems like the release process to public release is broken.

**Comment by AndrewGoodrich** [comment]
So that suggests the change from December 2021 was not migrated into 2.3.4.

**Comment by jakobs** [comment]
@AndrewGoodrich would it be possible that you could include the PR for OSCI-wg/systemc #435 in this public repository? It seems cleaner than replicating your changes based on the issue trail available.

**Comment by AndrewGoodrich** [comment]
I have made the changes in a copy of 2.3.4 in a sandbox on my machine, should I just do a push to get the changes into the repository?

Thanks,
   Andy

**Comment by jakobs** [comment]
> I have made the changes in a copy of 2.3.4 in a sandbox on my machine, should I just do a push to get the changes into the repository? Thanks, Andy
@AndrewGoodrich I don't know how this public repository relates to the OSCI-wg/systemc internal one, but I would guess the cleanest way would be for you to clone this public repository, push your changes to that clone and open a Pull Request in this repository. Also it would be great if somehow these changes could make it into an official release.
I much appreciate your support!

**Comment by AndrewGoodrich** [comment]
Its pull request #59

**Comment by chenhaoc** [comment]
There is a simple way to avoid `double free` when exit. That is to force sc_curr_simcontext=NULL before sc_elab_and_sim. 
btw: 3.0.0 has removed the global state `none`. 

```C++
if(cold_start) {
    sc_curr_simcontext = NULL; // force to be null before sc_elab
    ret = sc_elab_and_sim(argc,argv);
    cold_start = false;
}
else {
    delete sc_curr_simcontext;
    sc_curr_simcontext = NULL; 
    ret = sc_elab_and_sim(argc,argv);
}
```

**Comment by markfoodyburton** [comment]
Above PR has been merged in 3.0.0. Closing this issue.
